### PR TITLE
Fixed the issue of incorrect item position when prototypeItem is set in SliverReorderableList.

### LINKED ISSUE #142708: `SliverReorderableList` adds extra unwanted space when dragging an item, only when `prototypeItem` is provided
### Steps to reproduce

1. Create a `SliverReorderableList` in which the `prototypeItem` is exactly the same as (and has the exact dimensions of) the item built in `itemBuilder`
2. Start dragging any item of the list

### Expected results

The item being dragged should be kept in place and the space that is "behind" it should be kept unchanged.

### Actual results

The space "behind" the item being dragged increases.

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  @override
  MyHomePageState createState() => MyHomePageState();
}

class MyHomePageState extends State<MyHomePage> {
  final items = [for (int i = 0; i < 10; i++) 'Item $i'];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('SliverReorderableList Example'),
      ),
      body: CustomScrollView(
        slivers: [
          SliverReorderableList(
            itemBuilder: (context, index) {
              return ReorderableDragStartListener(
                key: ValueKey(items[index]),
                index: index,
                child: _Tile(
                  Text(items[index]),
                ),
              );
            },
            itemCount: items.length,
            onReorder: (oldIndex, newIndex) {
              setState(() {
                if (newIndex > oldIndex) {
                  newIndex -= 1;
                }
                final item = items.removeAt(oldIndex);
                items.insert(newIndex, item);
              });
            },
            // Comment `prototypeItem` to fix the issue
            prototypeItem: const _Tile(
              Text('Item 0'),
            ),
          ),
        ],
      ),
    );
  }
}

final class _Tile extends StatelessWidget {
  const _Tile(this.title);

  final Widget title;

  @override
  Widget build(BuildContext context) {
    return Material(
      child: ListTile(
        title: title,
        trailing: const Icon(Icons.drag_handle),
      ),
    );
  }
}

```
</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

**Actual results**

https://github.com/flutter/flutter/assets/4605213/b4ff44aa-4d49-44ae-97a0-97ce19b218bd

**Expected results**
The expected results were achieved by removing the `prototypeItem` parameter

https://github.com/flutter/flutter/assets/4605213/16adb90f-87e3-4c14-af58-6a59749e88d0

</details>

### Logs

_No response_

### Flutter Doctor output

The sample code was made and ran in DartPad stable channel (Flutter 3.16.9 Dart SDK 3.2.6).

### PR DESCRIPTION
**Pull request by yiiim**
Fixes #142708

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/test/widgets/reorderable_list_test.dart [hunk 1]
```diff
@@ -1483,6 +1483,51 @@ void main() {
...
+                      index: index,
+                      child: SizedBox(
+                        height: 100,
+                        child: Text('$index'),
+                      ),
+                    );
+                  },
+                  itemCount: 5,
+                  itemExtent: itemExtent,
+                  prototypeItem: prototypeItem,
```
Comment (Piinks) [inline packages/flutter/test/widgets/reorderable_list_test.dart:1507]: Hmm, this looks like a bug. You should not be able to use itemExtent and prototypeItem together.

Which one does it end up using?

We assert against this in other ScrollViews, like ListView.

File: packages/flutter/lib/src/widgets/reorderable_list.dart [hunk 2]
```diff
@@ -1163,10 +1139,15 @@ class _ReorderableItemState extends State<_ReorderableItem> {
...
-        : Offset.zero;
+  void updateForGap(int dragIndex, int gapIndex, double gapExtent, bool animate, bool reverse) {
+    late final Offset newTargetOffset;
+    if (gapIndex < dragIndex && index < dragIndex && index >= gapIndex) {
+      newTargetOffset = _extentOffset(reverse ? -gapExtent : gapExtent, _listState._scrollDirection);
+    } else if (gapIndex > dragIndex && index > dragIndex && index < gapIndex) {
+      newTargetOffset = _extentOffset(reverse ? gapExtent : -gapExtent, _listState._scrollDirection);
+    } else {
+      newTargetOffset = Offset.zero;
+    }
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/reorderable_list.dart:1150]: This is a bit hard to follow, can you split up some of the logic, or add comments to help understand what is happening here?

File: packages/flutter/lib/src/widgets/reorderable_list.dart [hunk 3]
```diff
@@ -1140,7 +1115,8 @@ class _ReorderableItemState extends State<_ReorderableItem> {
   @override
   Widget build(BuildContext context) {
     if (_dragging) {
-      return const SizedBox();
+      final Size size = _extentSize(_listState._dragInfo!.itemExtent, _listState._scrollDirection);
+      return SizedBox(width: size.width, height: size.height);
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/reorderable_list.dart:1119]: Nit:
```suggestion
      return SizedBox.fromSize(size);
```

File: packages/flutter/test/widgets/reorderable_list_test.dart (hunk 1 above)
Comment (yiiim) [reply packages/flutter/test/widgets/reorderable_list_test.dart:1507]: They will not exist at the same time. This is just to provide a convenient method for testing `prototypeItem` and `itemExtent` respectively.

**Review by Piinks — APPROVED** [review]
This is a really nice clean up. Thank you! LGTM
